passkey: authentication phase 1

justin-stephenson opened the pull request:
Add support for performing passkey authentication with SSSD.

Requires testing with:

*  FIDO2 compatible device
*  `setenforce 0` until https://bugzilla.redhat.com/show_bug.cgi?id=2151361 is resolved.
*  IPA build including initial passkey support: https://copr.fedorainfracloud.org/coprs/ipedrosa/passkey-auth/

ikerexxe reviewed (changes requested):
I have added some minor comments inline.

Moreover, while doing some basic testing I have discovered errors regarding the password prompt. If I set `pam_passkey_auth = true` in sssd.conf and I also set a password for a user, then when I try to su the password prompt appears:
```
$ su - user@ipa.test
Insert your passkey device, then press ENTER.
Enter PIN:
Password:
```

If I don't set a password for the user, then it only asks for the password:
```
$ su - iker@ipa.test
Password:
```

In my opinion, neither of the two cases should prompt for the password and they should only ask for the passkey authentication.

I'll provide additional feedback while I progress in my testing.

alexey-tikhonov commented on Makefile.am:
```diff
@@ -808,6 +808,7 @@ dist_noinst_HEADERS = \
     src/db/sysdb_ssh.h \
     src/db/sysdb_subid.h \
     src/db/sysdb_domain_resolution_order.h \
+    src/db/sysdb_passkey_user_verification.h \
```
Please, keep entire code related to 'passkey' RFE build time configurable.

alexey-tikhonov replied:
(Just in case: this applies to man pages, etc as well)

alexey-tikhonov replied:
And everything should be eventually packaged into new `sssd-passkey` sub-package.

justin-stephenson replied:
code and man pages should all be build time configurable now.

justin-stephenson commented on src/man/sssd.conf.5.xml:
```diff
@@ -1704,6 +1749,18 @@ pam_cert_verification = partial_chain
+                            wait for passkey_child to finish.
+                        </para>
+                        <para>
+                            Default: 10
```
Fixed.

justin-stephenson commented on src/man/sssd.conf.5.xml:
```diff
@@ -4346,6 +4403,48 @@ ldap_user_extra_attrs = phone:telephoneNumber
+                    <varlistentry>
+                        <term>touch_prompt</term>
+                        <listitem>
+                            <para>to change the message of the touch prompt</para></listitem>
```
Fixed.

justin-stephenson commented on src/man/sssd.conf.5.xml:
```diff
@@ -4346,6 +4403,48 @@ ldap_user_extra_attrs = phone:telephoneNumber
+                        </listitem>
+                    </varlistentry>
+                    <varlistentry>
+                        <term>interactive_prompt</term>
```
Done.

alexey-tikhonov replied:
Would it be confusing if strings are localized? (are they?)

justin-stephenson replied:
Default messages are localized, are man pages not?
~~~
#define DEFAULT_PASSKEY_PROMPT_INTERACTIVE _("Insert your Passkey device, then press ENTER.")
#define DEFAULT_PASSKEY_PROMPT_TOUCH _("Please touch the device.")
~~~

alexey-tikhonov replied:
Translation coverage of man page is much worse than of strings.
Besides, I'm not sure if translation tools will enforce the same translation in different places (in strings and in the man page) but hopefully "yes".
Do we quote default value of other  prompts in the man page?

justin-stephenson replied:
No we don't provide default value for other prompts, I went ahead and removed the default messages from the man pages.

justin-stephenson commented on src/util/authtok.h:
```diff
@@ -438,4 +438,34 @@ errno_t sss_authtok_get_oauth2(struct sss_auth_token *tok,
 
+/**
+ * @brief Returns a const string if the auth token is of type
+          SSS_AUTHTOK_TYPE_PASSKEY, otherwise it returns an error
```
Fixed.

justin-stephenson commented on src/responder/pam/pamsrv.h:
```diff
@@ -39,6 +39,13 @@ enum pam_initgroups_scheme {
+enum passkey_user_verification {
+    PAM_PASSKEY_VERIFICATION_ON,
+    PAM_PASSKEY_VERIFICATION_OFF,
+    PAM_PASSKEY_VERIFICATION_UNSET,
```
Fixed.

justin-stephenson commented on src/responder/pam/pamsrv_cmd.c:
```diff
@@ -100,6 +130,12 @@ static errno_t check_cert(TALLOC_CTX *mctx,
+                           struct tevent_context *ev,
+                           struct pam_ctx *pctx,
+                           struct pam_auth_req *preq,
+                           struct pam_data *pd);
```
Fixed.

justin-stephenson commented on src/responder/pam/pamsrv_cmd.c:
```diff
@@ -1478,6 +1514,436 @@ static errno_t check_cert(TALLOC_CTX *mctx,
+        DEBUG(SSSDBG_TRACE_FUNC, "Passkey verification is being enforced from local configuration\n");
+    }
+    DEBUG(SSSDBG_TRACE_FUNC, "Passkey verification setting [%s]\n",
+                             pam_passkey_verification_enum_to_string(verification));
```
No indentation issue here that I see.

justin-stephenson commented on src/responder/pam/pamsrv_passkey.c:
```diff
@@ -0,0 +1,369 @@
+                              tevent_req_data(req, struct pam_passkey_auth_send_state);
+    state->child_status = WEXITSTATUS(child_status);
+    if (WIFEXITED(child_status)) {
+            if (WEXITSTATUS(child_status) != 0) {
```
Fixed.

justin-stephenson commented on src/sss_client/pam_sss.c:
```diff
@@ -251,8 +252,7 @@ static int do_pam_conversation(pam_handle_t *pamh, const int msg_style,
     if (ret != PAM_SUCCESS) return ret;
-    if (conv == NULL || conv->conv == NULL) {
-        logger(pamh, LOG_ERR, "No conversation function");
+    if (conv == NULL || conv->conv == NULL) { logger(pamh, LOG_ERR, "No conversation function");
```
Fixed.

justin-stephenson commented on src/sss_client/pam_sss_prompt_config.c:
```diff
@@ -187,6 +222,8 @@ static errno_t pc_list_add_pc(struct prompt_config ***pc_list,
 #define DEFAULT_2FA_PROMPT_1ST _("First Factor: ")
 #define DEFAULT_2FA_PROMPT_2ND _("Second Factor: ")
+#define DEFAULT_PASSKEY_INTERACTIVE_PROMPT _("Insert your Passkey device, then press ENTER.")
+#define DEFAULT_PASSKEY_TOUCH_PROMPT _("Please touch the device. ")
```
Fixed, removed the duplicates.

ikerexxe commented:
> ```
> $ su - user@ipa.test
> Insert your passkey device, then press ENTER.
> Enter PIN:
> Password:
> ```

The password prompt is a fallback in case the PIN is not entered but it is required.

> If I don't set a password for the user, then it only asks for the password:
> 
> ```
> $ su - iker@ipa.test
> Password:
> ```

That was my error for not registering the key properly in the IPA server.

Apart from that, I have continued testing with IPA and everything seems good.

ikerexxe reviewed (commented):
I was able to test the functionality with a LDAP server and it works perfectly.

If you fix the issue that I discovered with the timeout then I think this PR should be ready.

justin-stephenson commented on src/util/util.h:
```diff
@@ -797,6 +797,10 @@ errno_t create_preauth_indicator(void);
 #define P11_CHILD_PATH SSSD_LIBEXEC_PATH"/p11_child"
 #define P11_CHILD_TIMEOUT_DEFAULT 10
 #define P11_WAIT_FOR_CARD_TIMEOUT_DEFAULT 60
+#define PASSKEY_CHILD_TIMEOUT_DEFAULT 10
```
Fixed, thank you.

ikerexxe commented on src/tests/cmocka/test_pam_srv.c:
```diff
@@ -4485,6 +4612,12 @@ int main(int argc, const char *argv[])
                                         pam_test_setup_2fa_single_and_service_prompt, pam_test_teardown),
         cmocka_unit_test_setup_teardown(test_pam_prompting_2fa_single_and_service_srv,
                                         pam_test_setup_2fa_single_and_service_prompt, pam_test_teardown),
+#ifdef BUILD_PASSKEY
+        cmocka_unit_test_setup_teardown(test_pam_prompting_passkey_interactive,
+                                        pam_test_setup_passkey_interactive_prompt, pam_test_teardown),
+        cmocka_unit_test_setup_teardown(test_pam_prompting_passkey_interactive_and_touch,
+                                        pam_test_setup_passkey_interactive_and_touch_prompt, pam_test_teardown),
+#endif
```
To keep it tidy, you could add the call to these tests along with the other passkey tests.

justin-stephenson replied:
Fixed.

ikerexxe reviewed (approved):
I have one last minor comment. Apart from that everything looks good to me. As the comment isn't important I'm approving the changes.

alexey-tikhonov commented on src/man/sssd.conf.5.xml:
```diff
@@ -722,6 +722,40 @@
+                                options. Supported options are:
+                                <variablelist>
+                                <varlistentry>
+                                    <term>user_verification=true|false</term>
```
Typically it's denoted as ` (boolean)` in SSSD man pages.

justin-stephenson replied:
Fixed.

alexey-tikhonov commented on src/man/sssd.conf.5.xml:
```diff
@@ -722,6 +722,40 @@
+                                    </listitem>
+                                </varlistentry>
+                                </variablelist>
+                            </para>
```
Mention a default value for `passkey_verification`?

alexey-tikhonov commented on src/man/sssd.conf.5.xml:
```diff
@@ -722,6 +722,40 @@
+                                    <term>debug_libfido2</term>
+                                    <listitem>
+                                        <para> Enable libfido2 debug
+                                        messages.</para>
```
Disabled by default?

alexey-tikhonov commented:
New config option should be mentioned in the release notes.
But perhaps entire feature will be covered by a dedicated single section in RNs?

alexey-tikhonov commented on src/man/sssd-ldap-attributes.5.xml:
```diff
@@ -636,6 +636,18 @@
+                            mapping data of the user.
+                        </para>
+                        <para>
+                            Default: passkey
```
Should it mention default value for LDAP/IPA/AD?

Like:
```
Default: homeDirectory (LDAP and IPA), unixHomeDirectory (AD)
```

justin-stephenson replied:
Fixed, thanks.

alexey-tikhonov commented on src/db/sysdb_passkey_user_verification.h:
```diff
@@ -0,0 +1,46 @@
+                                       struct ldb_dn *dn,
+                                       const char *user_verification);
+
+errno_t sysdb_domain_get_passkey_user_verification(TALLOC_CTX *mem_ctx,
```
Formatting feels inconsitent (return type on the same line with function name vs on its own line).

A brief comment describing functions purpose would help.

alexey-tikhonov commented on src/db/sysdb_passkey_user_verification.c:
```diff
@@ -0,0 +1,235 @@
+        goto done;
+    } else {
+        /* res->count == 1 */
+        attr = ldb_msg_find_attr_as_string(res->msgs[0], attrs[0], NULL);
```
What is the reason to have this additional helper - `sysdb_get_passkey_user_verification_string_attr()` - that supports multiple attr names, if code expects (and uses only) `attrs[0]`?

justin-stephenson replied:
You are right, I corrected this helper function to use SYSDB_PASSKEY_USER_VERIFICATION attr only.

alexey-tikhonov replied:
Thanks.
Frankly, I would question if this helper is needed at all, but as you see fit.

alexey-tikhonov commented on Makefile.am:
```diff
@@ -2597,6 +2600,7 @@ pam_srv_tests_SOURCES = \
     src/sss_client/pam_message.c \
     src/responder/pam/pamsrv_cmd.c \
     src/responder/pam/pamsrv_p11.c \
+    src/responder/pam/pamsrv_passkey.c \
```
`if BUILD_PASSKEY`?

alexey-tikhonov replied:
Ah, it's done in a later commit.

justin-stephenson commented:
> New config options should be mentioned in the release notes. But perhaps entire feature will be covered by a dedicated single section in RNs?

I will bring this as a topic to our next passkey meeting.

justin-stephenson commented:
> New config options should be mentioned in the release notes. But perhaps entire feature will be covered by a dedicated single section in RNs?

We do not plan to mention these new config options as part of this PR, we will advertise the feature, and related new options, when phase 2 completes. Certain aspects of this passkey implementation will also change as part of phase 2.

ikerexxe commented:
@madhuriupadhye has caught what might be a bug in these series of patches. The domain name for the authentication is set from the `domains` value in the sssd section of the configuration file, instead of the `*_domain` (ipa_domain, ad_domain, etc.) value in the domain section.

Example:
```
[sssd]
...
domains = ipa
...

[domain/ipa]
...
ipa_domain = ipa.test
...
```

In the previous example the domain name for the authentication is `ipa`, and not `ipa.test`.  Is that expected?

justin-stephenson commented:
> @madhuriupadhye has caught what might be a bug in these series of patches. The domain name for the authentication is set from the `domains` value in the sssd section of the configuration file, instead of the `*_domain` (ipa_domain, ad_domain, etc.) value in the domain section.
> 
> Example:
> 
> ```
> [sssd]
> ...
> domains = ipa
> ...
> 
> [domain/ipa]
> ...
> ipa_domain = ipa.test
> ...
> ```
> 
> In the previous example the domain name for the authentication is `ipa`, and not `ipa.test`. Is that expected?

I'm not sure what is best here, but FYI in phase 2 we will use the `domain` value returned from the the passkey plugin instead of the domain name returned from `struct cache_req_result`.

sumit-bose commented:
> > @madhuriupadhye has caught what might be a bug in these series of patches. The domain name for the authentication is set from the `domains` value in the sssd section of the configuration file, instead of the `*_domain` (ipa_domain, ad_domain, etc.) value in the domain section.
> > Example:
> > ```
> > [sssd]
> > ...
> > domains = ipa
> > ...
> > 
> > [domain/ipa]
> > ...
> > ipa_domain = ipa.test
> > ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In the previous example the domain name for the authentication is `ipa`, and not `ipa.test`. Is that expected?
> 
> I'm not sure what is best here, but FYI in phase 2 we will use the `domain` value returned from the the passkey plugin instead of the domain name returned from `struct cache_req_result`.

Hi,

if possible I would suggest to not use the domain name string from `sssd.conf`'s `[domain/...]` header because it can be any string and just has to match one of the strings from the `domains` list.

bye,
Sumit

justin-stephenson commented:
> the most easy way to solve this might be to add an additional `passkeyFallbackAuthDomain` or similar attribute to the user's cache entry if a `SYSDB_USER_PASSKEY` is written which contains the actual name of the domain the user is coming from.

Where would be a proper function/codepath to handle this? 

I am quite surprised `struct sss domain_info` does not contain this full domain name string already, should it be extended to do so? There must be other existing use cases where the full domain name is needed?

Thank you.

sumit-bose commented:
> > the most easy way to solve this might be to add an additional `passkeyFallbackAuthDomain` or similar attribute to the user's cache entry if a `SYSDB_USER_PASSKEY` is written which contains the actual name of the domain the user is coming from.
> 
> Where would be a proper function/codepath to handle this?
> 
> I am quite surprised `struct sss domain_info` does not contain this full domain name string already, should it be extended to do so? There must be other existing use cases where the full domain name is needed?

Hi,

adding the DNS domain name to `struct sss_domain_info` might be a good idea as well. There is already the `realm` member, which in most cases will be the same as the DNS domain name (only in upper-case). I guess you can add the DNS domain name in the same code paths where the `flat_name` (NETBios domain name) and `domain_id` (domain SID) members are added.

bye,
Sumit

> 
> Thank you.

justin-stephenson commented:
I added the commit [Add DNS Domain name to struct sss_domain_info](https://github.com/SSSD/sssd/pull/6491/commits/6709b7e0b47acff543369b34f9eabeb931bb3697) and switched the passkey code to use this dns domain name for the `--domain` passkey_child argument. If DNS domain name is `NULL` (LDAP Provider), fall back to the struct sss_domain_info `name` member.

justin-stephenson commented:
@ikerexxe some code in `check_passkey_verification` conditionally handles passkey configuration user verification setting as `default` in addition to `on/true` and `off/false`. Sorry I missed the discussion about this while on PTO, can this handling for `default` be removed?

ikerexxe commented:
> @ikerexxe some code in `check_passkey_verification` conditionally handles passkey configuration user verification setting as `default` in addition to `on/true` and `off/false`. Sorry I missed the discussion about this while on PTO, can this handling for `default` be removed?

If you are referring to the `else` branch in `check_passkey_verification()` (lines 293-295), then yes.

Sorry for not taking into account that you weren't in the meeting.

sumit-bose commented on src/providers/ipa/ipa_subdomains.c:
```diff
@@ -434,6 +435,12 @@ static errno_t ipa_subdom_store(struct sss_domain_info *parent,
         goto done;
     }
 
+    ret = sysdb_attrs_get_string(attrs, IPA_CN, &dns);
```
Hi,

I think it might be more clear to just copy `name` to make clear that it is the same for IPA (similar as with `realm` as upper-case version of `name`).

bye,
Sumit

justin-stephenson replied:
Fixed.

sumit-bose commented:
Hi,

thanks for the patches. I was able to run some successful tests. But if the user has multiple passkey entries in the LDAP server currently only one is used. So there is a loop over the configured entries needed somewhere. I'm not sure what would be the best place, in the PAM responder and call `passkey_child` multiple times or in `passkey_child`  which then most probably needs a new option to read key handle and public key from a single string so that multiple pairs can be given. @ikerexxe what do you think would be best?

As a general comment, I find the use of `#ifdef BUILD_PASSKEY` a bit excessive. I would prefer to only use it to protect areas where compilation would fail when `libfido2` is not available or to warn or error out explicitly e.g. when `pam_passkey_auth = True` is set but SSSD is build without passkey support. In all other areas, e.g. authtok, do not need  `#ifdef BUILD_PASSKEY`  imo.

bye,
Sumit

ikerexxe commented:
> thanks for the patches. I was able to run some successful tests. But if the user has multiple passkey entries in the LDAP server currently only one is used. So there is a loop over the configured entries needed somewhere. I'm not sure what would be the best place, in the PAM responder and call `passkey_child` multiple times or in `passkey_child` which then most probably needs a new option to read key handle and public key from a single string so that multiple pairs can be given. @ikerexxe what do you think would be best? 

The `passkey_child` is already able to handle the input of several key handles and public keys (comma separated). Example:
```
--key-handle=handle1,handle2 --public-key=key1,key2
```

Or are you suggesting to use another option to pass all the information as it comes from the server?

sumit-bose commented:
> > thanks for the patches. I was able to run some successful tests. But if the user has multiple passkey entries in the LDAP server currently only one is used. So there is a loop over the configured entries needed somewhere. I'm not sure what would be the best place, in the PAM responder and call `passkey_child` multiple times or in `passkey_child` which then most probably needs a new option to read key handle and public key from a single string so that multiple pairs can be given. @ikerexxe what do you think would be best?
> 
> The `passkey_child` is already able to handle the input of several key handles and public keys (comma separated). Example:
> 
> ```
> --key-handle=handle1,handle2 --public-key=key1,key2
> ```
> 
> Or are you suggesting to use another option to pass all the information as it comes from the server?

Hi,

thanks, yes, I was thinking of an option where the pair of handle and key are in a single string like they are stored in LDAP but since `passkey_child` can already multiple keys I think this is ok. So this PR should only be extended to send all the available key handles and public keys to `passkey_child`.

bye,
Sumit

justin-stephenson commented:
> Hi,
> 
> thanks for the patches. I was able to run some successful tests. But if the user has multiple passkey entries in the LDAP server currently only one is used. So there is a loop over the configured entries needed somewhere. I'm not sure what would be the best place, in the PAM responder and call `passkey_child` multiple times or in `passkey_child` which then most probably needs a new option to read key handle and public key from a single string so that multiple pairs can be given. @ikerexxe what do you think would be best?

Fixed, code now handles multiple passkey entries.
> 
> As a general comment, I find the use of `#ifdef BUILD_PASSKEY` a bit excessive. I would prefer to only use it to protect areas where compilation would fail when `libfido2` is not available or to warn or error out explicitly e.g. when `pam_passkey_auth = True` is set but SSSD is build without passkey support. In all other areas, e.g. authtok, do not need `#ifdef BUILD_PASSKEY` imo.

No actual libfido2 calls are made with my code, so I removed most of the `#ifdef BUILD_PASSKEY` lines except man pages and unit tests. I added a log message in the PAM responder in `may_do_passkey_auth` if passkey support is not built and to return false.

justin-stephenson commented:
latest push to this PR includes a fix for https://github.com/SSSD/sssd/issues/6558

The behavior was changed to have the PAM responder continue down the normal authentication path when passkey auth fails, or no passkey mapping is found for the user.

sumit-bose commented:
Hi,

please check to comments from Covscan and CodeQL.

bye,
Sumit

justin-stephenson commented:
> Hi,
> 
> please check to comments from Covscan and CodeQL.

CodeQL was reporting errors for code unrelated to this PR. I fixed some Covscan errors, a couple warnings remain but it looks like false positives. Please check.

justin-stephenson commented:
Latest push incorporates the following diff. This was needed to address an issue with AD/LDAP passkey authentication found by Madhuri.

~~~
diff --git a/src/responder/pam/pamsrv_passkey.c b/src/responder/pam/pamsrv_passkey.c
index 497f0e0ee..b466e8819 100644
--- a/src/responder/pam/pamsrv_passkey.c
+++ b/src/responder/pam/pamsrv_passkey.c
@@ -280,6 +280,8 @@ static errno_t check_passkey_verification(TALLOC_CTX *mem_ctx,
         DEBUG(SSSDBG_OP_FAILURE,
               "Failed to read passkeyUserVerification from sysdb: [%d]: %s\n",
               ret, sss_strerror(ret));
+        /* This is expected for AD and LDAP */
+        ret = EOK;
         goto done;
     }
 
@@ -476,7 +478,7 @@ void pam_passkey_get_user_done(struct tevent_req *req)
 
     /* Get passkey data */
     DEBUG(SSSDBG_TRACE_ALL, "Processing passkey data\n");
-    ret = process_passkey_data(pk_data, result->msgs[0], result->domain->dns_name, pk_data);
+    ret = process_passkey_data(pk_data, result->msgs[0], domain_name, pk_data);
     if (ret == ENOENT) {
         goto done;
     }
~~~

abbra commented on src/sss_client/pam_sss.c:
```diff
@@ -1776,6 +1790,121 @@ static int prompt_oauth2(pam_handle_t *pamh, struct pam_items *pi)
+        return PAM_SYSTEM_ERR;
+    }
+
+    pi->pam_authtok_type = SSS_AUTHTOK_TYPE_OAUTH2;
```
This should be SSS_AUTHTOK_TYPE_PASSKEY, right?

justin-stephenson replied:
correct, fixed.

abbra commented on src/sss_client/pam_sss.c:
```diff
@@ -1776,6 +1790,121 @@ static int prompt_oauth2(pam_handle_t *pamh, struct pam_items *pi)
+
+    pi->pam_authtok_type = SSS_AUTHTOK_TYPE_OAUTH2;
+	/* Interactive, prompt a message and wait before continuing */
+    if ((strcmp(prompt_interactive, "")) != 0) {
```
Not sure if `(prompt_interactive != NULL && prompt_interactive[0] != '\0')` would be better?

justin-stephenson replied:
Fixed.

abbra commented on src/sss_client/pam_sss.c:
```diff
@@ -1776,6 +1790,121 @@ static int prompt_oauth2(pam_handle_t *pamh, struct pam_items *pi)
+     *
+     * If prompt_pin is false but a PIN is set on the device
+     * we still prompt for PIN */
+    if ((strcmp(pi->passkey_prompt_pin, "true")) == 0) {
```
should this be case-insentive?

abbra commented on src/sss_client/pam_sss.c:
```diff
@@ -1776,6 +1790,121 @@ static int prompt_oauth2(pam_handle_t *pamh, struct pam_items *pi)
+    }
+
+    /* Prompt to remind the user to touch the device */
+    if ((strcmp(prompt_touch, "") != 0)) {
```
same here -- if you want to check that the string is not empty, `(foo != NULL && foo[0] != '\0')` is better than `strcmp()`

justin-stephenson replied:
Fixed.

abbra commented on src/sss_client/pam_sss.c:
```diff
@@ -1776,6 +1790,121 @@ static int prompt_oauth2(pam_handle_t *pamh, struct pam_items *pi)
+        return ret;
+    }
+
+    if ((strcmp(pi->passkey_prompt_pin, "true")) == 0) {
```
case-insensitive?

justin-stephenson replied:
Fixed.

abbra commented on src/sss_client/pam_sss.c:
```diff
@@ -1776,6 +1790,121 @@ static int prompt_oauth2(pam_handle_t *pamh, struct pam_items *pi)
+            }
+
+            /* Fallback to password auth if no PIN was entered */
+            if ((strcmp(resp[pin_idx].resp, "") == 0)) {
```
`(foo != NULL && foo[0] != '\0')` ?

justin-stephenson replied:
Fixed.

abbra reviewed (changes requested):
Few more comments to the passkey processing...

sumit-bose commented on src/responder/pam/pamsrv_passkey.c:
```diff
@@ -0,0 +1,923 @@
+            return EINVAL;
+        }
+
+        buf = talloc_size(mem_ctx, len);
```
Hi,

I would suggest to add `sss_erase_talloc_mem_securely()` as a talloc destructor to `buf` to make sure the pin is properly overwritten.

Additionally, the PIN is send without `\0` in the end which is completely ok. I found during testing that `passkey_child` might need some fix to make sure that it will not be confused about the length:
```
diff --git a/src/passkey_child/passkey_child_credentials.c b/src/passkey_child/passkey_child_credentials.c
index 9ebab97db..1eaf7bcc3 100644
--- a/src/passkey_child/passkey_child_credentials.c
+++ b/src/passkey_child/passkey_child_credentials.c
@@ -173,7 +173,7 @@ passkey_recv_pin(TALLOC_CTX *mem_ctx, int fd, char **_pin)
         return EINVAL;
     }
 
-    str = talloc_strdup(mem_ctx, (char *) buf);
+    str = talloc_strndup(mem_ctx, (char *) buf, len);
     if (str == NULL) {
         DEBUG(SSSDBG_CRIT_FAILURE, "talloc_strndup failed.\n");
         return ENOMEM;
```
`buf` is just a fixed size `uint8_t` array which is filled with the PIN. Since `buf` might contain random data the next byte following the received PIN might not be `\0` and as a result more bytes than expected will be copied into `str`. Using `talloc_strndup()` fixes this. I think it would be most straight forward to add this or a similar fix into your PR.

bye,
Sumit

justin-stephenson replied:
Thanks Sumit. I added a commit which adds the destructor for the PIN as you stated.

The other issue is already addressed in 7fb89ab01312bba9cb4099dc874725801f12bc78 if I am not wrong, I rebased this PR so it should be included now.

sumit-bose commented on src/responder/pam/pamsrv_passkey.c:
```diff
@@ -569,6 +569,17 @@ static void pam_passkey_auth_done(int child_status,
                                   struct tevent_signal *sige,
                                   void *pvt);
 
+int pin_destructor(void *ptr)
```
Hi,

please fix the comments about unused `ret` and incompatible types and I think you can make the destructor static as well.

As an unrelated comment, I realized that the PIN prompt `Enter PIN`  cannot be changed, I think a `pin_prompt` prompting option might be useful. But I#m fine if you prefer to add this with another PR.

bye,
Sumit

justin-stephenson replied:
Fixed.

Okay, I'll create a ticket for the `enter PIN` prompting option and address it separately.

sumit-bose reviewed (approved):
Hi,

thanks, ACK.

bye,
Sumit

abbra commented on src/responder/pam/pamsrv_passkey.c:
```diff
@@ -0,0 +1,935 @@
+    }
+
+    /* Use dns_name for AD/IPA - for LDAP fallback to domain->name */
+    domain_name = result->domain->dns_name;
```
covscan complains that there is no result->domain != NULL check. I think it wouldn't hurt to add one...

justin-stephenson replied:
Thanks Alexander, it is fixed now.

sumit-bose commented:
Hi,

I think this PR is ready to push. I checked that all of @abbra's comments are resolved and marked them accordingly.

bye,
Sumit

abbra commented:
I agree, I think this PR can be merged. I am a bit concerned about covscan warnings around packing of the passkey payload but it seems to be a confusion from the covscan side...

ikerexxe commented:
Moving back to `Changes requested` as the CI is failing due to a white-space issue. @justin-stephenson please take a look at it.

justin-stephenson commented:
> Moving back to `Changes requested` as the CI is failing due to a white-space issue. @justin-stephenson please take a look at it.

This is fixed now.

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/6491

* `master`
    * 5a22aefbc7acbafb2b0596eee0e9c668ea8e4b6d - PAM: Covscan NULL check for cache req result domain
    * 64aa2672e806bcf4631473d22b103a0369cc53ba - PAM: Add destructor for passkey pin
    * 938676a37e40a4345e25acd234ef79d25888ee8f - Tests: Passkey prompting config
    * 64f98463ac620b868d1ac8657bb9269707fc1f7d - Tests: Passkey (Pre)auth
    * 7f8fe399496d211d365f0054edf61fc611ec86ed - Support Passkey prompting config
    * b92ff263c2a918d727c016905959ed4bf7cb9c2c - PAM: Call the passkey helper binary
    * ea9bcab65c4c81fe36a43d6283f50eafcfe2b1c4 - IPA Retrieve passkey configuration
    * ec67749305ac41841d89c4808c9f9a637cc65a47 - Add DNS Domain name to struct sss_domain_info
    * 3f24aa71fa9dd26bf589275bb0d46775b6bf9b69 - Extend IPA config search
    * 57152761b23f9a5798c83861cc2948a20511ad5d - Add new option ldap_user_passkey
    * fbbe9ba3fbbd132abe0713587c367399c2e7ecd1 - authtok: Add Passkey type
    * fb5a300b493adb5a5534e68df23b6dd83bf92c4f - passkey: Add configuration options
